### ISSUE TITLE
Fatal error upgrading from `9.4.1` to `9.4.2` possibly relating to `register_ptk_patterns` and the WC_Logger

### ISSUE THREAD

**Issue by mikejolley** [woothemes/woocommerce#53042]
### Prerequisites

- [ ] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

A user on the forums reported a fatal error on upgrade. The stack trace indicates something relating to `register_ptk_patterns` and the logger. 

Forum post: https://wordpress.org/support/topic/upgrade-from-9-4-1-to-9-4-2-breaks-entire-site/

Maybe related to the changes that were cherry picked recently: https://github.com/woocommerce/woocommerce/pull/52725, however it looks more likely that the `WC_Logger` class is failing and this is causing a fatal error. `register_ptk_patterns` uses the logger if patterns are not found.

If `register_ptk_patterns` is not at fault, the logger may need some improvements to prevent fatal errors surfacing.

### Expected behavior

No errors on upgrade.

### Actual behavior

Fatal error that must be resolved by downgrading. This is the Stack Trace (I made GPT improve readability).

```
### PHP Fatal Error Log

**Date & Time:**  
`[Thu Nov 21 08:26:35.735260 2024]`  

**Error Details:**  
`PHP Fatal error: Uncaught TypeError: ftp_nlist(): Argument #1 ($ftp) must be of type FTP\Connection, null given`  

**File:**  
`/var/www/x/wp-admin/includes/class-wp-filesystem-ftpext.php`  
**Line:**  
`438`  

---

### Stack Trace
1. **Function Call:** `ftp_nlist()`  
   - **Location:** `/var/www/x/wp-admin/includes/class-wp-filesystem-ftpext.php` (Line 438)  
2. **Function Call:** `WP_Filesystem_FTPext->exists()`  
   - **Location:** `/var/www/x/wp-admin/includes/class-wp-filesystem-ftpext.php` (Line 456)  
3. **Function Call:** `WP_Filesystem_FTPext->is_file()`  
   - **Location:** `/var/www/x/wp-admin/includes/class-wp-filesystem-ftpext.php` (Line 265)  
4. **Function Call:** `Automattic\WooCommerce\Internal\Admin\Logging\FileV2\File->create()`  
   - **Location:** `/var/www/x/wp-content/plugins/woocommerce/src/Internal/Admin/Logging/FileV2/File.php` (Line 417)  
5. **Function Call:** `Automattic\WooCommerce\Internal\Admin\Logging\FileV2\File->write()`  
   - **Location:** `/var/www/x/wp-content/plugins/woocommerce/src/Internal/Admin/Logging/FileV2/File.php` (Line 438)  
6. **Function Call:** `Automattic\WooCommerce\Internal\Admin\Logging\FileV2\FileController->write_to_file()`  
   - **Location:** `/var/www/x/wp-content/plugins/woocommerce/src/Internal/Admin/Logging/FileV2/FileController.php` (Line 135)  
7. **Function Call:** `Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2->handle()`  
   - **Location:** `/var/www/x/wp-content/plugins/woocommerce/src/Internal/Admin/Logging/LogHandlerFileV2.php` (Line 60)  
8. **Function Call:** `WC_Logger->log()`  
   - **Location:** `/var/www/x/wp-content/plugins/woocommerce/includes/class-wc-logger.php` (Line 189)  
9. **Function Call:** `WC_Logger->warning()`  
   - **Location:** `/var/www/x/wp-content/plugins/woocommerce/includes/class-wc-logger.php` (Line 268)  
10. **Function Call:** `Automattic\WooCommerce\Blocks\BlockPatterns->register_ptk_patterns()`  
    - **Location:** `/var/www/x/wp-content/plugins/woocommerce/src/Blocks/BlockPatterns.php` (Line 164)  
11. **Function Call:** `WP_Hook->apply_filters()`  
    - **Location:** `/var/www/x/wp-includes/class-wp-hook.php` (Line 324)  
12. **Function Call:** `WP_Hook->do_action()`  
    - **Location:** `/var/www/x/wp-includes/class-wp-hook.php` (Line 348)  
13. **Function Call:** `do_action()`  
    - **Location:** `/var/www/x/wp-includes/plugin.php` (Line 517)  

---

### Context
- **WordPress Config File:** `/var/www/x/wp-config.php` (Line 94)  
- **WordPress Load File:** `/var/www/x/wp-load.php` (Line 50)  
- **WordPress Blog Header File:** `/var/www/x/wp-blog-header.php` (Line 13)  
- **Index File:** `/var/www/x/index.php` (Line 17)  

---
```

### Steps to reproduce

There are no steps to reproduce aside from upgrading, but this does not impact all stores.

### WordPress Environment

N/A

### Isolating the problem

- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [ ] I can reproduce this bug consistently using the steps above.

**Comment by gigitux** [comment]
https://github.com/woocommerce/woocommerce/pull/52725 has been deployed with WooCommerce 9.4.1, so I don't think that it is the culprit. In WooCommerce 9.4.2, we shipped the following PRs:

- https://github.com/woocommerce/woocommerce/pull/52900
- https://github.com/woocommerce/woocommerce/pull/51540
- https://github.com/woocommerce/woocommerce/pull/51631

I checked these PRs and it looks like that none of these PRs should be responsible for this regression. However, it would be prudent to conduct further investigation and improve the logger to prevent such fatal errors if needed. @woocommerce/proton, what do you think? Is it something that you can work on?

**Comment by coreymckrill** [comment]
It does seem like #51540 is the only change that could be related. However, I don't see a way that this could happen, because if WP_Filesystem does not succeed in initializing (which is what would happen if the FTP connection fails) then it wouldn't get as far as it did to cause the fatal error. Unfortunately setting up a test environment where the local file system is locked down and WP uses FTP instead is exceedingly difficult. It would be interesting to know if this particular site was able to record logs before this crash happened.

Of note, there is a related upstream issue in WordPress Core where the potential for this fatal error was identified, but [they chose not to fix it](https://core.trac.wordpress.org/ticket/58940#comment:15).

**Comment by linuxchris12** [comment]
Hi, I can provide more logs, if needed.

I can also try to apply 9.4.1 with just this [PR](https://github.com/woocommerce/woocommerce/pull/51540) on the site if you think that makes sense.

**Comment by linuxchris12** [comment]
Now I played around a bit more with it and applied the 3 PRs manually to my WP installation - that did not lead to any error.
I then applied 9.4.2 manually directly, and it is working as well.

So I can't reproduce the issue anymore. Something must have gone wrong during the original update process, maybe a file was not written or got corrupted or similar or there were some old left-over files in the woocommerce dir which caused problems (the woocommerce installation is from 2016).
So a full deletion of the woocommerce plugin folder and re-creation of it from the official sources fixed the issue. Sorry for the trouble.

This can be closed.
